### ISSUE TITLE
Introduce some practical experience in ant group.

### ISSUE THREAD

**Issue by Lingtaonju** [quicwg/load-balancers#119]
Dear  Duke and Banks,
as you know , in ant group, the biggest fintech company in the world, we have designed and run the quic lb to support quic migration and other characteristic for many years.  We are glad to see that you  are trying to make quic lb standard.  
In addition to  the content of this onging quic lb draft,  I want to provide some more practical experience in ant group which maybe make it more substantial.

In this issue, I will  firstly show you the stateless reset  in quic lb, details here: 

```
As the transport-draft described, stateless reset provided as an option of last resort for an
endpoint that does not have access to the state of a connection. To realize this, endpoint(e.g: server) 
sends stateless reset packet to the sender(such as client). The stateless reset packet is stateless, 
so, in the cluster, if all the servers use the same static key, it will work.
More details about how is omitted here. 

```

As reset is stateless,  we make  it  work in quic lb. 
```
1. quic lb use the same  encryption algorithm (HKDF, HMACK) and static key  with quic server.
2. for valid quic short header packet with server info encoded, but corresponding ups is down 
   or removed from the cluster, quic lb will generate the valid reset packet to make client terminate 
   the connection.
3. other details  I wii show you in the PR
```

About the stateless, we can discuss here. Looking forward to your feedback

**Comment by martinduke** [comment]
Interesting design --- thanks for the PR!

I will have to think about this some more, but here are some quick reactions, mostly related to security. None of these mean that I'm rejecting the idea.

(1) QUIC-LB is already very complicated (maybe too complicated?), and I am reluctant to add yet more features. Can you share data on how this helps in your deployment? To me, stateless reset is an edge case that wouldn't happen all that often and didn't have particularly important benefits to offloading it.

In particular, this implies an out-of-band method to communicate server state. If a server was going down for maintenance or at a rate-limit, why wouldn't the server simply close connections and refuse new ones, or use its own flow control? Why not use a Retry service to redirect new connections to a different server?

(2) The consequences of leaking this key are especially large -- for SCID/BCID, it's just compromising the server mapping; for Retry Services, you can break the Retry mechanism. For this, an attacker can inject packets to kill any connection it can observe.

(3) There probably needs to be some further protection against Stateless Reset Oracle (Sec 21.11 of quic-transport). I believe it would be sufficient to only allow this in response to short header packets.

(4) The configuration actions need much more definition to be a usable standard. There needs to be a format of the information you're encrypting, in addition to a key and cipher.

(5) Unlike the Retry Services piece, this directly ties in with load balancing, but load balancing tries to be version-independent and this not version independent.

(6) This can't work without length self-encoding; otherwise, the LB can't reliably extract the CID.

**Comment by Lingtaonju** [comment]
Dear duke, thanks for your reply. For these considerations, I made some following additions，we can discuss further :
+ For question (1)

```
“QUIC-LB is already very complicated (maybe too complicated?), and I am reluctant to add yet more features.”
```

I agree with this view partially. In the design of a certain function, it is complicated and many things are considered，but in terms of the features of the Quic LB itself, I think more can be added?  QUIC LB can do more things for this complicated UDP protocol, more features means necessity but not confusion ?

```
" To me, stateless reset is an edge case that wouldn't happen all that often and didn't have particularly important benefits to offloading it"
```
The necessity of stateless rest in quic lb can be summarized:
1. Take the common open source based access network ( client -> lvs -> nginx -> real server)  for example, lvs has the ability to reset unknown tcp packets, in other words, lvs has tcp stack ability.  QUIC-LB  Maybe also needs this ability?  QUIC LB should reset the unkown QUIC packets, but not just drop it. So this is the basic ability,  although it may not wok often, it is essential.
 
2. Consider a typical situation where the backend carried in the DCID does not exist or unhealty because of maintenance. This situation is similar to that tcp's session state does not exist in LVS. It should be reset directly by QUIC LB instead of drop or send to other machines ？
 
```
"In particular, this implies an out-of-band method to communicate server state. If a server was going down for maintenance or at a rate-limit, why wouldn't the server simply close connections and refuse new ones, or use its own flow control? Why not use a Retry service to redirect new connections to a different server"
```
Rate-limit description in PR is not  particularly suitable, as stateless reset is not used to cancel initial packet, but rate-limit almost works in connection created, so I will delete this in PR. 
This reset works for connection already established while server had restarted or other reason lost the state,  so the server  can reset like tcp of cause,  the same to other servers in the cluster, but as stated above I think QUIC LB  should also this ability.

+ For question (2)(3)(4)(6) ,  I think more design detailes can be discussed later ?  Just like firewall and lb's tcp reset can bring security risks, quic lb's reset may also not avoid this risk , but can be designed more carefully.

+ For question(5),  I'm sorry that I  didn't get the point, why reset is version-related, but not retry service?  I think neither reset nor retry, quic version in lb and server needs match. 

Waiting for you reply.

**Comment by martinduke** [comment]
> 
> I agree with this view partially. In the design of a certain function, it is complicated and many things are considered，but in terms of the features of the Quic LB itself, I think more can be added?  QUIC LB can do more things for this complicated UDP protocol, more features means necessity but not confusion ?
> 

At the next meeting, I will ask if there is interest in addressing this use case.

> ```
> " To me, stateless reset is an edge case that wouldn't happen all that often and didn't have particularly important benefits to offloading it"
> ```
> 
> The necessity of stateless rest in quic lb can be summarized:
> 
>     1. Take the common open source based access network ( client -> lvs -> nginx -> real server)  for example, lvs has the ability to reset unknown tcp packets, in other words, lvs has tcp stack ability.  QUIC-LB  Maybe also needs this ability?  QUIC LB should reset the unkown QUIC packets, but not just drop it. So this is the basic ability,  although it may not wok often, it is essential.
> 
>     2. Consider a typical situation where the backend carried in the DCID does not exist or unhealty because of maintenance. This situation is similar to that tcp's session state does not exist in LVS. It should be reset directly by QUIC LB instead of drop or send to other machines?

I don't think "DCID [I think you mean server ID?] does not exist" is a real use case, unless the packet is garbage or an attack, as clients should always use server-generated CIDs in any short header. QUIC-LB can be implemented statelessly, so there is no analogy to the LB not having session state. In fact, if there is no record of a 4-tuple, this generally indicates a NAT rebinding or an address migration, and the connection should definitely NOT be reset.

If a server is going down for maintenance, one would think the operator would make an effort to close down existing sessions gracefully (perhaps through HTTP/3 goaway). If the LB simply stops treating that as a valid server ID, there won't be more connections routed there.

> This reset works for connection already established while server had restarted or other reason lost the state, so the server can reset like tcp of cause, the same to other servers in the cluster, but as stated above I think QUIC LB should also this ability.
> 
>     * For question (2)(3)(4)(6) ,  I think more design detailes can be discussed later ?  Just like firewall and lb's tcp reset can bring security risks, quic lb's reset may also not avoid this risk , but can be designed more carefully.
> 
>     * For question(5),  I'm sorry that I  didn't get the point, why reset is version-related, but not retry service?  I think neither reset nor retry, quic version in lb and server needs match.

Retry services are version-dependent, but they act on Initial Packets that encode the version. Stateless Resets operation on 1-RTT packets with no version information.

Furthermore, while in principle Retry Services can be the same device as the LB, I doubt that will be the case in practice. (I've considered splitting Retry Services into a separate draft for this reason). However, as a "Reset Service" would need to understand the state of the server pool, and would usually be part of the load balancer instance. It would therefore introduce version-dependence to a device that is currently roughly version-independent.

**Comment by Lingtaonju** [comment]
As blocked by something, sorry for the late reply.

**Firstly :**
> At the next meeting, I will ask if there is interest in addressing this use case.

Great, thanks. I am looking forward to the upcoming discussion.

**Secondly:**
> Retry services are version-dependent, but they act on Initial Packets that encode the version. Stateless Resets operation on 1-RTT packets with no version information.

But, I think this is the same problem for quic server who should send stateless reset? As this short header packet's version is also unknown for quic server. So, stateless reset should be version-independent?

**Thirdly:**
> However, as a "Reset Service" would need to understand the state of the server pool, and would usually be part of the load balancer instance.

I tend to think this should be lb's inner function, but not lb's third-party service. This function may be not used often, but should not missing.

**fourthly:**
> I don't think "DCID [I think you mean server ID?] does not exist" is a real use case, unless the packet is garbage or an attack, as clients should always use server-generated CIDs in any short header.

Yes, I mean DCID is SID here which encoded server info. But my point is , "the backend or server " carried in the DCID , doesn't appear in quic_lb's upstream pool. This is the key point for the necessary to introduce reset in quic lb.
As for the reason why the server info does not exist in lb's upstream pool, this may have many causes, for example the backend is deleted because of unhealty.

> If a server is going down for maintenance, one would think the operator would make an effort to close down existing sessions gracefully (perhaps through HTTP/3 goaway)

This is the graceful method, but for some interupt/ accidents case, reset also needs for quiclb  the same as quic server do.

**Comment by martinduke** [comment]
As I presented at IETF 112, I am leaning towards splitting up the draft into separate LB and Retry documents, as these functions are quite separate. Assuming that this split occurs, I think it makes sense to propose Reset Offload as a separate function and draft.

**Comment by Lingtaonju** [comment]
> 

Nice,  in this case,  more offload func can be added.
Looking forward to the next news

**Comment by martinduke** [comment]
Closing -- stateless reset is out of scope for either draft.
